sources/constraint-declaration-validation.asciidoc

```diff
@@ -1573,7 +1930,184 @@ The resolutions shown in <<table-constraintvalidator-resolution>> occur.
+
+Exactly one value extractor must be identified when processing a container element constraint or container element marked with `@Valid`.
+
+[[constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-registering]]
+===== Registering ValueExtractor implementations
+
+Value extractors can be registered with the validation engine in the following ways (in increasing order of priority):
+
+* Provided by the validation engine itself (see <<constraintdeclarationvalidationprocess-valueextractordefinitions-builtinvalueextractors>>)
+* Via the Java https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html[service loader] mechanism; for this the file _META-INF/services/javax.validation.valueextraction.ValueExtractor_ must be provided,
+with the fully-qualified name(s) of one or more extractor implementations as its contents
+* By specifying the fully-qualified class name of one or several extractors in `META-INF/validation.xml` (see <<validationapi-bootstrapping-xmlconfiguration>>)
+* By invoking the method `Configuration#addValueExtractor(ValueExtractor<?>)` (to apply it at the validation factory level)
+* By invoking the method `ValidatorContext#addValueExtractor(ValueExtractor<?>)` (to apply it for a single `Validator`)
+
+A value extractor for a given type and type parameter specified at a higher level overrides any other extractors for the same type and type parameter given at lower levels.
+If e.g. a value extractor defined as `class MyListValueExtractor implements ValueExtractor<List<@ExtractedValue ?>> { ... }` is given via `ValidatorContext#addValueExtractor(ValueExtractor<?>)`,
+it will take precedence over any other value extractor implementing `List<@ExtractedValue ?>` given via `Configuration#addValueExtractor(ValueExtractor<?>)`,
+_META-INF/validation.xml_ or the service loader mechanism as well as the built-in extractor for `List` values.
+
+[[constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-constraints]]
+===== ValueExtractor resolution algorithm for container element constraints
+
+For a container with the declared type `C` whose element type is hosting a constraint, the following resolution rules apply for identifying the value extractor:
+
+* A `ValueExtractor<T>` is said to be _type-compliant_ with `C`, if `C` is a subtype of `T`
+(according to the http://java.sun.com/docs/books/jls/third_edition/html/typesValues.html#4.10[Java Language Specification 3rd edition chapter 4.10 Subtyping]).
+Note that `C` is a subtype of `T` if `C` = `T`.
+* A `ValueExtractor` implementation is said to be _container-element-compliant_ with `C`, if `C` is a generic container type and the value extractor implementation handles a type parameter that maps to the constrained type argument.
+* If no `ValueExtractor` type-compliant and container-element-compliant with `C` is found among the available value extractors, a `ConstraintDeclarationException` is raised.
```

Comment:
Interesting, you've thought this through and an exception is better than ignoring? Just asking here to make sure we are not missing anything.